### DISCUSSION TITLE
Should mypy raise errors on missing kwargs for __init__ on factory functions?

### DISCUSSION THREAD

**Discussion by fernandobperezm** [python/typing#1191] [General]
I have been trying to create a _factory function_ for my classes and subclasses. 

Most subclasses have common `args` and `kwargs`. However, my main problem is that some subclasses have `kwargs` that are unique for them. 

An MRE of my situation is the following:

```python
from typing import TypeVar, Type


class A:
    def __init__(self, b: int):
        ...


class B(A):
    def __init__(self, b: int, c: str):
        ...


class C(A):
    def __init__(self, b: int, d: int):
        ...


T = TypeVar('T', bound=A)


def foo(cls: Type[T], *args, **kwargs) -> T:
    return cls(*args, **kwargs)


foo(B)
foo(B, b=1, c="1")

foo(C)
foo(C, b=1, d=1)
```

I'm wondering if a type checker (specifically looking at mypy) should raise errors for both `foo(B)` and `foo(C)` since I'm not passing the required `kwargs`.  Is this something that mypy (or another type-checker) is able to do?

My reasoning is that I want to be able to call that function and them mypy tell me if I'm missing any `arg` or `kwarg`. I know that I can get the errors by running the interpreter, however, when dealing with several kwargs, it can be time-consuming to execute the run the interpreter several times just to know that I missed a kwarg.

NOTE: This is my first discussion/question in this repo so I don't know if this is the best place to ask this question or even if it should be asked here. I very much welcome suggestions/corrections about this topic too 😊. Thanks!

**Comment by erictraut** [comment]
You can use `ParamSpec` (see PEP 612 for details).

Here's how that would be specified in your example use case.

```python
P = ParamSpec("P")
T = TypeVar("T", bound=A)

def foo(cls: Callable[P, T], *args: P.args, **kwargs: P.kwargs) -> T:
    return cls(*args, **kwargs)
```

**Reply by fernandobperezm** [reply]
Thanks! I see that `ParamSpec` is currently available in Python 3.10. I'm at Python 3.9, so I'm using typing_extensions, it should work in the same way, right?. 

mypy is currently raising the following errors:
- Name "P.args" is not defined
- Name "P.kwargs" is not defined
- The first argument to Callable must be a list of types or "..."

And the snippet that I sent now is:

```python
from typing import TypeVar, Callable
from typing_extensions import ParamSpec


class A:
    def __init__(self, b: int):
        ...


class B(A):
    def __init__(self, b: int, c: str):
        ...


class C(A):
    def __init__(self, b: int, d: int):
        ...


P = ParamSpec("P")
T = TypeVar('T', bound=A)


def foo(cls: Callable[P, T], *args: P.args, **kwargs: P.kwargs) -> T:
    return cls(*args, **kwargs)


foo(B)
foo(B, b=1, c="1")

foo(C)
foo(C, b=1, d=1)
```

Idk if it is a bug in mypy or I'm missing something else?

Thanks for your quick reply!

**Reply by AlexWaygood** [reply]
What version of mypy are you using? This should be supported on the latest version of mypy.

**Reply by fernandobperezm** [reply]
Oh my, I'm sorry, I ran update before replying and it told me I was updated (prob sending an issue now to Poetry 😂 as I was not). 

I was running mypy 0.910. Updated to 0.950 and now it works!

Thanks!!

**Reply by AlexWaygood** [reply]
Awesome -- no worries!

**Comment by fernandobperezm** [comment]
If I try to use this same approach on a `classmethod` decorated method inside A, mypy throws the following error: `Mypy: The erased type of self "def (*P.args, **P.kwargs) -> T'-2" is not a supertype of its class "Type[types.A]"`.

To avoid repetitions, the only change is in the definition of A:
```python
class A:
    def __init__(self, b: int):
        ...

    @classmethod
    def build(cls: Callable[P, T], *args: P.args, **kwargs: P.kwargs) -> T:
        return cls(*args, **kwargs)
```
I tried changing the function definition here and there but ended up in my original situation (so no error from mypy when args where missing) or with the error I just showed.

The semantics of this `build` function would be exactly the same as `foo` in my example. I would be use it in this way so subclasses have this `build` function defined.

**Reply by erictraut** [reply]
Yeah, mypy isn't happy with this usage, but it looks to me like it's valid. You might want to file a bug in the mypy tracker.
